rama-http/src/service/web/router.rs

```diff
@@ -0,0 +1,440 @@
+            prefix: prefix.to_owned(),
+            nested: service.into_endpoint_service().boxed(),
+        };
+
+        self.add_route(&path, matcher, nested_router_service)
+    }
+
+    /// add a route to the router with it's matcher and service.
+    fn add_route<I, T>(mut self, path: &str, matcher: HttpMatcher<State, Body>, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let service = service.into_endpoint_service().boxed();
+
+        if let Ok(matched) = self.routes.at_mut(path) {
+            matched.value.push((matcher, service));
+        } else {
+            self.routes
+                .insert(path, vec![(matcher, service)])
+                .expect("Failed to add route");
+        }
+
+        self
+    }
+
+    /// use the provided service when no route matches the request.
+    pub fn not_found<I, T>(mut self, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.not_found = Some(service.into_endpoint_service().boxed());
+        self
+    }
+}
+
+struct NestedRouterService<State> {
+    prefix: String,
+    nested: BoxService<State, Request, Response, Infallible>,
+}
+
+impl<State: std::fmt::Debug> std::fmt::Debug for NestedRouterService<State> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NestedRouterService")
+            .field("prefix", &self.prefix)
+            .finish()
+    }
+}
+
+impl<State> Service<State, Request> for NestedRouterService<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    type Response = Response;
+    type Error = Infallible;
+
+    async fn serve(
+        &self,
+        ctx: Context<State>,
+        mut req: Request,
+    ) -> Result<Self::Response, Self::Error> {
+        let path = ctx.get::<UriParams>().unwrap().glob().unwrap();
+
+        if path.starts_with(&self.prefix) {
+            // strip the "/prefix" from the request URI
+            let new_path = &path[self.prefix.len()..];
+            // update the request URI with the new path
+            *req.uri_mut() = new_path.parse().unwrap();
+        }
+
+        self.nested.serve(ctx, req).await
+    }
+}
+
+impl<State> Default for Router<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<State> Service<State, Request> for Router<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    type Response = Response;
+    type Error = Infallible;
+
+    async fn serve(
+        &self,
+        mut ctx: Context<State>,
+        req: Request,
+    ) -> Result<Self::Response, Self::Error> {
+        let mut ext = Extensions::new();
+
+        if let Ok(matched) = self.routes.at(req.uri().path()) {
+            for (matcher, service) in matched.value.iter() {
+                if matcher.matches(Some(&mut ext), &ctx, &req) {
+                    let uri_params = matched.params.iter().collect::<UriParams>();
```

Comment:
You might want to invent a new private type `ParentUriParams(UriParams)` that you do not expose but that you inject into the context from within `NestedService`. This way you can merge it here into the uri_params, in case it exists.

Reply:
Also make sure to add a test case for this, so it is validated to work, because while it is probably a niche use case I bet someone will eventually have parameters in their prefix :P